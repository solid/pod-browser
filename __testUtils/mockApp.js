/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  createThing,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, schema } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/addressBook";

export const oidcPrefix = "http://www.w3.org/ns/solid/oidc#";
export const TOS_PREDICATE = `${oidcPrefix}tos_uri`;
export const POLICY_PREDICATE = `${oidcPrefix}policy_uri`;
export const CONTACTS_PREDICATE = `${oidcPrefix}contacts`;
export const LOGO_PREDICATE = `${oidcPrefix}logo_uri`;

export const APP_WEBID = "https://mockappurl.com";
export const APP_POLICY_URL = "https://mockappurl.com/privacy-policy";
export const APP_TOS_URL = "https://mockappurl.com/terms-of-service";

const contactThing = chain(
  createThing({ url: "https://mockappurl.com#contacts" }),
  (t) => addStringNoLocale(t, vcard.fn, "Example Contact"),
  (t) => addStringNoLocale(t, vcard.hasTelephone, "555-5555"),
  (t) => addStringNoLocale(t, vcard.hasEmail, "example@email.com"),
  (t) => addUrl(t, vcard.url, "https://examplewebsite.com")
);

const dataset = setThing(createSolidDataset(), contactThing);

export function mockApp() {
  return chain(
    createThing({ url: APP_WEBID }),
    (t) => addStringNoLocale(t, vcard.fn, "Mock App"),
    (t) => addUrl(t, vcardExtras("WebId"), APP_WEBID),
    (t) => addUrl(t, rdf.type, schema.SoftwareApplication), // just guessing what the type might be for apps
    (t) => addUrl(t, TOS_PREDICATE, APP_TOS_URL),
    (t) => addUrl(t, POLICY_PREDICATE, APP_POLICY_URL),
    (t) => addUrl(t, CONTACTS_PREDICATE, "https://mockappurl.com#contacts")
  );
}

export function mockAppDataset() {
  return setThing(dataset, mockApp());
}
