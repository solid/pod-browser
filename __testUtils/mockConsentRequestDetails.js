/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

async function getConsentRequestDetails() {
  return {
    "@context": [
      "https://www.w3.org/2018/credentials/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1",
      "https://consent.pod.inrupt.com/credentials/v1",
    ],
    id: "https://consent.pod.inrupt.com/vc/53973727-f4d0-9e8dbdc041fd",
    type: ["VerifiableCredential", "SolidCredential", "SolidConsentRequest"],
    issuer: "https://consent.pod.inrupt.com",
    issuanceDate: "2021-05-26T16:40:03Z",
    expirationDate: "2021-06-23T16:40:03Z",
    credentialSubject: {
      id: "https://pod.inrupt.com/bob/profile/card#me",
      inbox: "https://pod.inrupt.com/bob/inbox/",
      hasConsent: {
        mode: ["Read", "Write"],
        hasStatus: "ConsentStatusRequested",
        forPersonalData: "https://pod.inrupt.com/alice/private/data",
        forPurpose: "https://example.com/SomeSpecificPurpose",
      },
    },
    proof: {
      created: "2021-05-26T16:40:03.009Z",
      proofPurpose: "assertionMethod",
      proofValue: "eqp8h_kL1DwJCpn65z-d1Arnysx6b11...jb8j0MxUCc1uDQ",
      type: "Ed25519Signature2020",
      verificationMethod: "https://consent.pod.inrupt.com/key/396f686b4ec",
    },
  };
}

export default getConsentRequestDetails;
