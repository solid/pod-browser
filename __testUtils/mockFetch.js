/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export default function mockFetch(responses = {}) {
  return jest.fn().mockImplementation((url, init) => {
    const responseFn = responses[url.toString()];
    const responseObject = responseFn;
    const method = init?.method || "GET";
    let response;
    if (responseFn && typeof responseObject[method] !== "undefined") {
      response = responseObject[method](url, init);
    } else if (responseFn) {
      response = responseFn(url, init);
    } else {
      throw new Error(`URL (${url}) not mocked properly`);
    }
    Object.defineProperty(response, "url", { value: url });
    return Promise.resolve(response);
  });
}

export function mockSolidClientRequest(body, method = "GET", headers = {}) {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status, body = "") {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
