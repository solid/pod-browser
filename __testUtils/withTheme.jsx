/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import T from "prop-types";
import { StylesProvider, ThemeProvider } from "@inrupt/prism-react-components";
import { FeatureProvider } from "../src/contexts/featureFlagsContext";
import defaultTheme from "../src/theme";

export default function WithTheme(props) {
  const { theme, children } = props;
  return (
    <StylesProvider generateClassName={(rule) => `PodBrowser-${rule.key}`}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </StylesProvider>
  );
}

WithTheme.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  theme: T.object,
  children: T.node.isRequired,
};

WithTheme.defaultProps = {
  theme: defaultTheme,
};

export const renderWithTheme = (children, theme = defaultTheme) => {
  return render(
    <FeatureProvider>
      <WithTheme theme={theme}>{children}</WithTheme>
    </FeatureProvider>
  );
};
