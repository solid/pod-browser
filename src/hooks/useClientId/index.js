/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useEffect, useState } from "react";

export default function useClientId(providerIri) {
  const [response, setResponse] = useState(null);
  const [error, setError] = useState(null);
  useEffect(() => {
    const checkOidcSupport = async () => {
      try {
        const res = await fetch(
          `${providerIri}/.well-known/openid-configuration`
        );
        const { solid_oidc_supported: solidOidcSupported } = await res.json();
        setResponse(
          solidOidcSupported &&
            solidOidcSupported === "https://solidproject.org/TR/solid-oidc"
        );
      } catch {
        setError("Fetching failed");
      }
    };
    checkOidcSupport();
  }, [providerIri]);
  return { response, error };
}
