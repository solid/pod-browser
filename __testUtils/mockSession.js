/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import mockFetch from "./mockFetch";
import mockResponse from "./mockResponse";

import profileTurtle from "./mocks/profile.ttl";
import storageTurtle from "./mocks/podRoot.ttl";
import storageAclTurtle from "./mocks/storageAcl.ttl";

export const webIdUrl = "http://example.com/webId#me";
export const storageUrl = "http://example.com/";
export const storageAclUrl = "http://example.com/.acl";
export const anotherUsersStorageUrl = "http://anotheruser.com/";
export const anotherUsersStorageAclUrl = "http://anotheruser.com/.acl";

export { profileTurtle };

export default function mockSession(options = {}) {
  // this instance is a normal, authenticated session
  // you can override the properties with what you need, but if you repeat something a lot
  // consider naming it and create a reusable mock function
  const { webId, ...restOptions } = options;
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webId || webIdUrl]: () =>
        mockResponse(200, profileTurtle, { "Content-Type": "text/turtle" }),
      [storageUrl]: () =>
        mockResponse(200, storageTurtle, {
          Link: `<${storageAclUrl}>; rel="acl"`,
          "Content-Type": "text/turtle",
        }),
      [storageAclUrl]: () =>
        mockResponse(200, storageAclTurtle, { "Content-Type": "text/turtle" }),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: options.webId || webIdUrl,
    }, // add more properties as needed
    ...restOptions,
  }; // add more properties as needed
}

// create other functions in this file to reflect other types of sessions we need to test with

export function mockUnauthenticatedSession() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: () => mockFetch(),
    info: {
      isLoggedIn: false,
      sessionId: null,
      webId: null,
    },
    login: jest.fn(),
  };
}

export function mockAuthenticatedSessionWithNoAccessToPod() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webIdUrl]: () => mockResponse(200, profileTurtle),
      [storageUrl]: () =>
        mockResponse(200, "", {
          Link: `<${storageAclUrl}>; rel="acl"`,
        }),
      [storageAclUrl]: () => mockResponse(401),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: webIdUrl,
    },
  };
}

export function mockAuthenticatedSessionWithNoAccessToAnotherUsersPod() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webIdUrl]: () => mockResponse(200, profileTurtle),
      [anotherUsersStorageUrl]: () =>
        mockResponse(200, "", {
          Link: `<${anotherUsersStorageAclUrl}>; rel="acl"`,
        }),
      [anotherUsersStorageAclUrl]: () => mockResponse(401),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: webIdUrl,
    },
  };
}

export const mockAuthenticatedSession =
  mockAuthenticatedSessionWithNoAccessToPod;
