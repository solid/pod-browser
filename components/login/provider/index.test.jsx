/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { waitFor } from "@testing-library/dom";
import { act } from "@testing-library/react-hooks";
import userEvent from "@testing-library/user-event";
import { fireEvent } from "@testing-library/react";
import { mockUnauthenticatedSession } from "../../../__testUtils/mockSession";
import mockSessionContextProvider from "../../../__testUtils/mockSessionContextProvider";
import ProviderLogin, {
  getErrorMessage,
  setupErrorHandler,
  setupLoginHandler,
  setupOnProviderChange,
  TESTCAFE_ID_LOGIN_FIELD,
} from "./index";
import * as useClientId from "../../../src/hooks/useClientId";
import { renderWithTheme } from "../../../__testUtils/withTheme";
import useIdpFromQuery from "../../../src/hooks/useIdpFromQuery";

jest.mock("../../../src/windowHelpers");
jest.mock("../../../src/hooks/useIdpFromQuery");

describe("ProviderLogin form", () => {
  beforeEach(() => {
    useIdpFromQuery.mockReturnValue(null);
  });

  it("renders a webid login form", () => {
    const { asFragment } = renderWithTheme(<ProviderLogin />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls the login function upon submitting the form", async () => {
    const session = mockUnauthenticatedSession();
    const SessionProvider = mockSessionContextProvider(session);
    const { login } = session;
    const { getByTestId } = renderWithTheme(
      <SessionProvider>
        <ProviderLogin />
      </SessionProvider>
    );
    const loginInput = getByTestId(TESTCAFE_ID_LOGIN_FIELD);
    userEvent.type(loginInput, "{enter}");
    waitFor(() => {
      expect(login).toHaveBeenCalled();
    });
  });

  it("renders a validation error if login fails", () => {
    const { asFragment } = renderWithTheme(
      <ProviderLogin defaultError={new Error()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("allows setting idp with query param", () => {
    const iri = "http://example.com";
    const label = "example.com";
    useIdpFromQuery.mockReturnValue({
      iri,
      label,
    });
    const { getByTestId } = renderWithTheme(<ProviderLogin />);
    const input = getByTestId(TESTCAFE_ID_LOGIN_FIELD).querySelector("input");
    expect(input.value).toEqual(label);
    expect(document.activeElement).toEqual(input);
  });

  it("calls checkOidcSupported method with the correct value on change", async () => {
    const iri = "http://example.com";
    const label = "example.com";
    useIdpFromQuery.mockReturnValue({
      iri,
      label,
    });

    const checkOidcSupportSpy = jest
      .spyOn(useClientId, "checkOidcSupport")
      .mockImplementation(async () => true);
    const { getByTestId } = renderWithTheme(<ProviderLogin />);
    const input = getByTestId(TESTCAFE_ID_LOGIN_FIELD).querySelector("input");
    expect(input.value).toEqual(label);

    fireEvent.change(input, {
      target: { value: "https://broker.pod.inrupt.com" },
    });
    expect(checkOidcSupportSpy).toHaveBeenCalledWith(
      "https://broker.pod.inrupt.com"
    );
  });
});

describe("setupOnProviderChange", () => {
  it("sets up event handler", () => {
    const setProviderIri = jest.fn();
    const setLoginError = jest.fn();
    setupOnProviderChange(setProviderIri, setLoginError)({}, "string");
    expect(setLoginError).toHaveBeenCalledWith(null);
    expect(setProviderIri).toHaveBeenCalledWith("https://string");
  });
  it("calls setProviderIri with provided string if user providers a correct URL", () => {
    const setProviderIri = jest.fn();
    const setLoginError = jest.fn();
    setupOnProviderChange(setProviderIri, setLoginError)({}, "https://string");
    expect(setLoginError).toHaveBeenCalledWith(null);
    expect(setProviderIri).toHaveBeenCalledWith("https://string");
  });
  it("calls setProviderIri with correct iri when passed an object from the autocomplete options", () => {
    const setProviderIri = jest.fn();
    const setLoginError = jest.fn();
    setupOnProviderChange(setProviderIri, setLoginError)(
      {},
      { iri: "https://example.com", label: "example.com" }
    );
    expect(setLoginError).toHaveBeenCalledWith(null);
    expect(setProviderIri).toHaveBeenCalledWith("https://example.com");
  });
  it("calls setProviderIri with null for other values", () => {
    const setProviderIri = jest.fn();
    const setLoginError = jest.fn();
    setupOnProviderChange(setProviderIri, setLoginError)({}, 42);
    expect(setLoginError).toHaveBeenCalledWith(null);
    expect(setProviderIri).toHaveBeenCalledWith(null);
  });
});

describe("setupLoginHandler", () => {
  it("sets up event handler", () => {
    const login = jest.fn();
    const event = { preventDefault: jest.fn() };
    setupLoginHandler(login)(event);
    expect(event.preventDefault).toHaveBeenCalledWith();
    expect(login).toHaveBeenCalledWith();
  });
});

describe("setupErrorHandler", () => {
  it("sets up event handler", () => {
    const setLoginError = jest.fn();
    const error = new Error();
    setupErrorHandler(setLoginError)(error);
    expect(setLoginError).toHaveBeenCalledWith(error);
  });
});

describe("getErrorMessage", () => {
  it("has a standard message", () =>
    expect(getErrorMessage(new Error())).toEqual(
      "We were unable to log in with this URL. Please fill out a valid Solid Identity Provider."
    ));
  it("handles when URL is not an IdP for Chrome, Edge, and Firefox", () =>
    expect(getErrorMessage(new Error("fetch"))).toEqual(
      "This URL is not a Solid Identity Provider."
    ));
  it("handles when URL is not an IDP for Safari", () =>
    expect(
      getErrorMessage(new Error("Not allowed to request resource"))
    ).toEqual("This URL is not a Solid Identity Provider."));
  it("handles when value is empty", () =>
    expect(getErrorMessage(new Error("sessionId"))).toEqual(
      "Please fill out a valid Solid Identity Provider."
    ));
});
