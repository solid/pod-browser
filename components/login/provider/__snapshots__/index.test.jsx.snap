// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProviderLogin form renders a validation error if login fails 1`] = `
<DocumentFragment>
  <form
    class="PodBrowser-provider-login__form"
  >
    <div
      class="PodBrowser-provider-login__wrapper"
    >
      <div
        class="PodBrowser-root PodBrowser-selectionBox"
      >
        <div
          aria-expanded="false"
          class="PodBrowser-root PodBrowser-hasClearIcon"
          role="combobox"
        >
          <div
            aria-describedby="login-error-text"
            class="PodBrowser-root PodBrowser-root PodBrowser-fullWidth"
            data-testid="login-field"
          >
            <div
              class="PodBrowser-root PodBrowser-root PodBrowser-inputRoot PodBrowser-error PodBrowser-error PodBrowser-fullWidth PodBrowser-formControl PodBrowser-adornedEnd PodBrowser-adornedEnd"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="true"
                autocapitalize="none"
                autocomplete="off"
                class="PodBrowser-input PodBrowser-input PodBrowser-input PodBrowser-inputFocused PodBrowser-inputAdornedEnd PodBrowser-inputAdornedEnd"
                id="provider-select"
                spellcheck="false"
                type="url"
                value=""
              />
              <div
                class="PodBrowser-endAdornment"
              >
                <button
                  aria-label="Clear"
                  class="PodBrowser-root PodBrowser-root PodBrowser-clearIndicator PodBrowser-clearIndicatorDirty"
                  tabindex="-1"
                  title="Clear"
                  type="button"
                >
                  <span
                    class="PodBrowser-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="PodBrowser-root PodBrowser-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    class="PodBrowser-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="PodBrowser-root PodBrowser-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PodBrowser-legend"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <p
          class="PodBrowser-root PodBrowser-error"
          id="login-error-text"
          style="margin-bottom: 10px;"
        >
          We were unable to log in with this URL. Please fill out a valid Solid Identity Provider.
        </p>
      </div>
      <div
        role="button"
        tabindex="0"
      >
        <button
          class="PodBrowser-button PodBrowser-provider-login__button"
          data-testid="go-button"
          type="submit"
        >
          <span
            class="PodBrowser-button__text"
          >
            Go
          </span>
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`ProviderLogin form renders a webid login form 1`] = `
<DocumentFragment>
  <form
    class="PodBrowser-provider-login__form"
  >
    <div
      class="PodBrowser-provider-login__wrapper"
    >
      <div
        class="PodBrowser-root PodBrowser-selectionBox"
      >
        <div
          aria-expanded="false"
          class="PodBrowser-root PodBrowser-hasClearIcon"
          role="combobox"
        >
          <div
            class="PodBrowser-root PodBrowser-root PodBrowser-fullWidth"
            data-testid="login-field"
          >
            <div
              class="PodBrowser-root PodBrowser-root PodBrowser-inputRoot PodBrowser-fullWidth PodBrowser-formControl PodBrowser-adornedEnd PodBrowser-adornedEnd"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="PodBrowser-input PodBrowser-input PodBrowser-input PodBrowser-inputFocused PodBrowser-inputAdornedEnd PodBrowser-inputAdornedEnd"
                id="provider-select"
                spellcheck="false"
                type="url"
                value=""
              />
              <div
                class="PodBrowser-endAdornment"
              >
                <button
                  aria-label="Clear"
                  class="PodBrowser-root PodBrowser-root PodBrowser-clearIndicator PodBrowser-clearIndicatorDirty"
                  tabindex="-1"
                  title="Clear"
                  type="button"
                >
                  <span
                    class="PodBrowser-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="PodBrowser-root PodBrowser-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    class="PodBrowser-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="PodBrowser-root PodBrowser-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PodBrowser-legend"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        role="button"
        tabindex="0"
      >
        <button
          class="PodBrowser-button PodBrowser-provider-login__button"
          data-testid="go-button"
          type="submit"
        >
          <span
            class="PodBrowser-button__text"
          >
            Go
          </span>
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`ProviderLogin form renders a webid login form with a pre-populated input if idp is available from query 1`] = `
<DocumentFragment>
  <form
    class="PodBrowser-provider-login__form"
  >
    <div
      class="PodBrowser-provider-login__wrapper"
    >
      <div
        class="PodBrowser-root PodBrowser-selectionBox"
      >
        <div
          aria-expanded="false"
          class="PodBrowser-root PodBrowser-focused PodBrowser-hasClearIcon"
          role="combobox"
        >
          <div
            class="PodBrowser-root PodBrowser-root PodBrowser-fullWidth"
            data-testid="login-field"
          >
            <div
              class="PodBrowser-root PodBrowser-root PodBrowser-inputRoot PodBrowser-fullWidth PodBrowser-focused PodBrowser-focused PodBrowser-formControl PodBrowser-adornedEnd PodBrowser-adornedEnd"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="PodBrowser-input PodBrowser-input PodBrowser-input PodBrowser-inputFocused PodBrowser-inputAdornedEnd PodBrowser-inputAdornedEnd"
                id="provider-select"
                spellcheck="false"
                type="url"
                value="http://example.com"
              />
              <div
                class="PodBrowser-endAdornment"
              >
                <button
                  aria-label="Clear"
                  class="PodBrowser-root PodBrowser-root PodBrowser-clearIndicator PodBrowser-clearIndicatorDirty"
                  tabindex="-1"
                  title="Clear"
                  type="button"
                >
                  <span
                    class="PodBrowser-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="PodBrowser-root PodBrowser-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    class="PodBrowser-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="PodBrowser-root PodBrowser-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PodBrowser-legend"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        role="button"
        tabindex="0"
      >
        <button
          class="PodBrowser-button PodBrowser-provider-login__button"
          data-testid="go-button"
          type="submit"
        >
          <span
            class="PodBrowser-button__text"
          >
            Go
          </span>
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
